From edef628381c7506fd688f2f2486e11aa890aacab Mon Sep 17 00:00:00 2001
From: Abhinav Kumar <quic_abhinavk@quicinc.com>
Date: Wed, 27 Aug 2025 15:32:30 +0800
Subject: [PATCH 1/2] FROMLIST: drm/msm/dp: Add support for sending VCPF
 packets in DP controller

The VC Payload Fill (VCPF) sequence is inserted by the DP controller
when stream symbols are absent, typically before a stream is disabled.
This patch adds support for triggering the VCPF sequence in the MSM DP
controller.

Signed-off-by: Abhinav Kumar <quic_abhinavk@quicinc.com>
Signed-off-by: Yongxing Mou <yongxing.mou@oss.qualcomm.com>
Signed-off-by: bowehan <bowehan@qti.qualcomm.com>
Link: https://lore.kernel.org/all/20250825-msm-dp-mst-v3-21-01faacfcdedd@oss.qualcomm.com/
Upstream-Status: Submitted [https://lore.kernel.org/all/20250825-msm-dp-mst-v3-21-01faacfcdedd@oss.qualcomm.com/]
---
 drivers/gpu/drm/msm/dp/dp_ctrl.c    | 42 +++++++++--------------------
 drivers/gpu/drm/msm/dp/dp_ctrl.h    |  3 +--
 drivers/gpu/drm/msm/dp/dp_display.c |  5 +---
 3 files changed, 14 insertions(+), 36 deletions(-)

diff --git a/drivers/gpu/drm/msm/dp/dp_ctrl.c b/drivers/gpu/drm/msm/dp/dp_ctrl.c
index f1238d8bb04d..df43017c689e 100644
--- a/drivers/gpu/drm/msm/dp/dp_ctrl.c
+++ b/drivers/gpu/drm/msm/dp/dp_ctrl.c
@@ -133,37 +133,18 @@ static int msm_dp_aux_link_configure(struct drm_dp_aux *aux,
 	return 0;
 }
 
-void msm_dp_ctrl_push_idle(struct msm_dp_ctrl *msm_dp_ctrl)
+void msm_dp_ctrl_push_idle(struct msm_dp_ctrl *msm_dp_ctrl, struct msm_dp_panel *msm_dp_panel)
 {
 	struct msm_dp_ctrl_private *ctrl;
-
-	ctrl = container_of(msm_dp_ctrl, struct msm_dp_ctrl_private, msm_dp_ctrl);
-
-	reinit_completion(&ctrl->idle_comp);
-	msm_dp_catalog_ctrl_state_ctrl(ctrl->catalog, DP_STATE_CTRL_PUSH_IDLE);
-
-	if (!wait_for_completion_timeout(&ctrl->idle_comp,
-			IDLE_PATTERN_COMPLETION_TIMEOUT_JIFFIES))
-		pr_warn("PUSH_IDLE pattern timedout\n");
-
-	drm_dbg_dp(ctrl->drm_dev, "mainlink off\n");
-}
-
-void msm_dp_ctrl_push_vcpf(struct msm_dp_ctrl *msm_dp_ctrl, struct msm_dp_panel *msm_dp_panel)
-{
 	u32 state = 0x0;
-	struct msm_dp_ctrl_private *ctrl;
 
 	ctrl = container_of(msm_dp_ctrl, struct msm_dp_ctrl_private, msm_dp_ctrl);
 
-	if (msm_dp_panel->stream_id >= DP_STREAM_MAX) {
-		DRM_ERROR("invalid input\n");
-		return;
-	}
-
-	if (msm_dp_panel->stream_id == DP_STREAM_0)
+	if (!ctrl->mst_active)
+		state |= DP_STATE_CTRL_PUSH_IDLE;
+	else if (msm_dp_panel->stream_id == DP_STREAM_0)
 		state |= MST_DP0_PUSH_VCPF;
-	else
+	else if (msm_dp_panel->stream_id == DP_STREAM_1)
 		state |= MST_DP1_PUSH_VCPF;
 
 	reinit_completion(&ctrl->idle_comp);
@@ -171,8 +152,8 @@ void msm_dp_ctrl_push_vcpf(struct msm_dp_ctrl *msm_dp_ctrl, struct msm_dp_panel
 	msm_dp_catalog_ctrl_state_ctrl(ctrl->catalog, state);
 
 	if (!wait_for_completion_timeout(&ctrl->idle_comp,
-					 IDLE_PATTERN_COMPLETION_TIMEOUT_JIFFIES))
-		pr_warn("PUSH_VCPF pattern timedout\n");
+			IDLE_PATTERN_COMPLETION_TIMEOUT_JIFFIES))
+		pr_warn("PUSH_IDLE pattern timedout\n");
 
 	drm_dbg_dp(ctrl->drm_dev, "mainlink off\n");
 }
@@ -1598,7 +1579,7 @@ void msm_dp_ctrl_set_psr(struct msm_dp_ctrl *msm_dp_ctrl, bool enter)
 			return;
 		}
 
-		msm_dp_ctrl_push_idle(msm_dp_ctrl);
+		msm_dp_ctrl_push_idle(msm_dp_ctrl, ctrl->panel);
 		msm_dp_catalog_ctrl_state_ctrl(ctrl->catalog, 0);
 
 		msm_dp_catalog_ctrl_psr_mainlink_enable(ctrl->catalog, false);
@@ -1708,7 +1689,7 @@ int msm_dp_ctrl_mst_send_act(struct msm_dp_ctrl *msm_dp_ctrl)
 	act_complete = msm_dp_catalog_read_act_complete_sts(ctrl->catalog);
 
 	if (!act_complete)
-		DRM_ERROR("mst ACT trigger complete SUCCESS\n");
+		drm_dbg_dp(ctrl->drm_dev, "mst ACT trigger complete SUCCESS\n");
 	else
 		drm_dbg_dp(ctrl->drm_dev, "mst ACT trigger complete failed\n");
 
@@ -1720,7 +1701,7 @@ static int msm_dp_ctrl_link_maintenance(struct msm_dp_ctrl_private *ctrl)
 	int ret = 0;
 	int training_step = DP_TRAINING_NONE;
 
-	msm_dp_ctrl_push_idle(&ctrl->msm_dp_ctrl);
+	msm_dp_ctrl_push_idle(&ctrl->msm_dp_ctrl, ctrl->panel);
 
 	ctrl->link->phy_params.p_level = 0;
 	ctrl->link->phy_params.v_level = 0;
@@ -2302,6 +2283,7 @@ void msm_dp_ctrl_off_link(struct msm_dp_ctrl *msm_dp_ctrl)
 	msm_dp_catalog_ctrl_mainlink_ctrl(ctrl->catalog, false);
 	msm_dp_catalog_mst_config(ctrl->catalog, false);
 
+	msm_dp_catalog_ctrl_reset(ctrl->catalog);
 	ctrl->mst_active = false;
 
 	dev_pm_opp_set_rate(ctrl->dev, 0);
@@ -2453,7 +2435,7 @@ static int msm_dp_ctrl_clk_init(struct msm_dp_ctrl *msm_dp_ctrl)
 
 	ctrl->pixel_clk[DP_STREAM_1] = devm_clk_get(dev, "stream_1_pixel");
 	if (IS_ERR(ctrl->pixel_clk[DP_STREAM_1]))
-		DRM_ERROR("failed to get stream_1_pixel clock");
+		DRM_DEBUG("failed to get stream_1_pixel clock");
 	else
 		ctrl->num_pixel_clks++;
 
diff --git a/drivers/gpu/drm/msm/dp/dp_ctrl.h b/drivers/gpu/drm/msm/dp/dp_ctrl.h
index f72a499809ec..7f45000161a2 100644
--- a/drivers/gpu/drm/msm/dp/dp_ctrl.h
+++ b/drivers/gpu/drm/msm/dp/dp_ctrl.h
@@ -24,7 +24,7 @@ int msm_dp_ctrl_prepare_stream_on(struct msm_dp_ctrl *dp_ctrl, bool force_link_t
 void msm_dp_ctrl_off_link_stream(struct msm_dp_ctrl *msm_dp_ctrl);
 void msm_dp_ctrl_off_link(struct msm_dp_ctrl *msm_dp_ctrl);
 void msm_dp_ctrl_stream_clk_off(struct msm_dp_ctrl *msm_dp_ctrl, struct msm_dp_panel *msm_dp_panel);
-void msm_dp_ctrl_push_idle(struct msm_dp_ctrl *msm_dp_ctrl);
+void msm_dp_ctrl_push_idle(struct msm_dp_ctrl *msm_dp_ctrl, struct msm_dp_panel *msm_dp_panel);
 irqreturn_t msm_dp_ctrl_isr(struct msm_dp_ctrl *msm_dp_ctrl);
 void msm_dp_ctrl_handle_sink_request(struct msm_dp_ctrl *msm_dp_ctrl);
 struct msm_dp_ctrl *msm_dp_ctrl_get(struct device *dev, struct msm_dp_link *link,
@@ -52,7 +52,6 @@ void msm_dp_ctrl_mst_stream_channel_slot_setup(struct msm_dp_ctrl *msm_dp_ctrl,
 void msm_dp_ctrl_set_mst_channel_info(struct msm_dp_ctrl *msm_dp_ctrl,
 				      enum msm_dp_stream_id strm,
 				      u32 start_slot, u32 tot_slots);
-void msm_dp_ctrl_push_vcpf(struct msm_dp_ctrl *dp_ctrl, struct msm_dp_panel *msm_dp_panel);
 
 int msm_dp_ctrl_get_stream_cnt(struct msm_dp_ctrl *dp_ctrl);
 
diff --git a/drivers/gpu/drm/msm/dp/dp_display.c b/drivers/gpu/drm/msm/dp/dp_display.c
index 25da8d8de82d..5ac957010843 100644
--- a/drivers/gpu/drm/msm/dp/dp_display.c
+++ b/drivers/gpu/drm/msm/dp/dp_display.c
@@ -1630,10 +1630,7 @@ void msm_dp_display_disable_helper(struct msm_dp *dp, struct msm_dp_panel *msm_d
 		return;
 	}
 
-	if (msm_dp_get_mst_max_stream(dp) > DEFAULT_STREAM_COUNT)
-		msm_dp_ctrl_push_vcpf(msm_dp_display->ctrl, msm_dp_panel);
-	else
-		msm_dp_ctrl_push_idle(msm_dp_display->ctrl);
+	msm_dp_ctrl_push_idle(msm_dp_display->ctrl, msm_dp_panel);
 
 	if (msm_dp_get_mst_max_stream(dp) > DEFAULT_STREAM_COUNT) {
 		msm_dp_ctrl_mst_stream_channel_slot_setup(msm_dp_display->ctrl,
-- 
2.34.1

